WIP: PBS-to-Slurm adaptation (#306)
hpcleuven/VscDocumentation: source/leuven/credits.rst

=======================================

KU Leuven uses a credit system to do accounting on the HPC systems it hosts.
This is the case for its Tier-2 clusters.

Information on how to use the credit system can be found in the :ref:`documentation

> on credit system basics <credit system basics>`.

Author: Thanks for bringing this under my radar. I would remove `credit_system_basics` because the `slurm-accounting` page, as you correctly mentioned, fully covers the same topic.

**CPU-only jobs**

The following formula applies::

   CPU TRESBillingWeights * num_cores * walltime

Where

- ``CPU TRESBillingWeights`` is the applied weight for CPU resources (see above)
- ``num_cores`` is the *effective* number of cores used for the job
- ``walltime`` is the number of minutes that the job ran

**GPU jobs**

The following formula applies::

   (CPU TRESBillingWeights * num_cores + GPU TRESBillingWeights * num_gpus) * walltime
   
Where

- ``CPU TRESBillingWeights`` is the applied weight for CPU resources (see above)
- ``GPU TRESBillingWeights`` is the applied weight for GPU resources (see above)
- ``num_cores`` is the *effective* number of cores used for the job
- ``num_gpus`` is the number of GPUs requested for the job
- ``walltime`` is the number of minutes that the job ran

.. note::

    *Effective* number of cores is not necessarily equal to what the user requests.
    E.g. if a job requests a single core/task, but the full memory of a node on wICE,
    then one node is blocked for such a job. Consequently, the effective number of cores
    will be 72, instead of 1.

    Similarly, on a node with 4 GPUs, if a user requests to use a single GPU, one-fourth

>     of cores will be assigned to the job. So, e.g. on wICE ``num_cores`` will be 18.

Reviewer: This stil needs to be removed

Author: frankly, everytime I tried to test this on wICE, the gpu nodes were busy. But, judging from `slurm.conf` no cpu-to-gpu proportion is set. Also on Slurm documentation, I hardly found a possibility to enforce this. there is only `DefCpuPerGPU` (https://slurm.schedmd.com/slurm.conf.html#OPT_DefCpuPerGPU) which only sets a default value, in case users do not specify it explicitly.

So, I believe we need to enforce this via the submit filter, same as now with PBS on Genius.

So, I will remove that paragraph now.

it consumes. Two distinct TRES are the number of CPU cores and GPU devices. 
Different types of CPU and GPU nodes are given different weights 
(``TRESBillingWeights``) which you can retrieve as follows for e.g. wICE::

   scontrol show partitions -M wice

As an example, for a GPU node, the weights are configured as::

   TRESBillingWeights=CPU=2.546296296,GRES/gpu:a100-sxm4-80gb=141.6666667,GRES/shard:a100-sxm4-80gb=141.6666667

**CPU-only jobs**

The following formula applies::

   CPU TRESBillingWeights * num_cores * walltime

Where

- ``CPU TRESBillingWeights`` is the applied weight for CPU resources (see above)
- ``num_cores`` is the *effective* number of cores used for the job
- ``walltime`` is the number of minutes that the job ran

**GPU jobs**

The following formula applies::

   (CPU TRESBillingWeights * num_cores + GPU TRESBillingWeights * num_gpus) * walltime
   
Where

- ``CPU TRESBillingWeights`` is the applied weight for CPU resources (see above)
- ``GPU TRESBillingWeights`` is the applied weight for GPU resources (see above)
- ``num_cores`` is the *effective* number of cores used for the job

> - ``num_gpus`` is the *effective* number of GPUs used for the job

Reviewer: ```suggestion
- ``num_gpus`` is the number of GPUs requested for the job
```
(see further explanation below)

- ``walltime`` is the number of minutes that the job ran

.. note::

    *Effective* number of cores is not necessarily equal to what the user requests.
    E.g. if a job requests a single core/task, but the full memory of a node on wICE,
    then one node is blocked for such a job. Consequently, the effective number of cores
    will be 72, instead of 1.

>     Similarly, on a node with 4 GPUs, if a user requests to use a single GPU, one-fourth
>     of cores will be assigned to the job. So, e.g. on wICE ``num_cores`` will be 18.
> 
> .. note::
> 
>     *Effective* number of GPUs deserves an explanation. If a user requests e.g. 2 GPUs
>     for a job, then ``num_gpus`` will be 2. However, in certain cases, we allow multiple
>     users to share a GPU (called sharding). In such cases, ``num_gpus`` could be a factor
>     between zero and one.

Reviewer: ```suggestion
```
We have not (yet) implemented a mechanism in our Slurm setup that would ensure that a certain cores-to-GPU ratio is applied in GPU jobs (e.g. 18:1 in the case of wICE). So it is best to just describe it as it is (i.e. without the notion of "effective" number of GPUs). The description of sharding is also not correct/complete here, and I suggest to just leave it out (at least for now), since this is only supposed to be used for specific courses, and definitely not in general.

Author: fair enough. I adapted the text, and also removed the corresponding "note" that elaborated the "effective gpus".

.. note::

    The Tier-2 cluster has several types of compute nodes.

>     Hence, different ``TRESBillingWeights`` applies to 

Reviewer: ```suggestion
    Hence, different ``TRESBillingWeights`` apply to 
```

    different resources on different partitions of Genius and wICE.
    The difference in cost between different machines/processors reflects
    the performance difference between those types of nodes.
    For additional information, you may refer to the 

>     `HPC Service Catalog<https://icts.kuleuven.be/sc/onderzoeksgegevens/hpc_vsc_page>`_

Author: YES

    (login required).

The difference in cost between different machines/processors reflects
the price-performance difference between those types of nodes. The total cost
of a job will be comparable on any compute node, but the
walltime will be different, depending on the performance of the nodes.

In the examples below, you run your jobs on a ``skylake`` node, for which
we charge 10 000 Slurm credits per hour.

An example of a job running on multiple nodes and cores is given below::

>    $ sbatch -A lp_astrophysics_014 -M genius -N 2 -ntasks-per-node 36 simulation_3415.slurm

Reviewer: ```suggestion
   $ sbatch -A lp_astrophysics_014 -M genius -N 2 --ntasks-per-node=36 simulation_3415.slurm
```

For Genius thin nodes we have ``TRESBillingWeights=CPU=4.62963``.
If this job finishes in 2.5 hours (i.e., walltime is 150 minutes), the user
will be charged::

   4.62963 * (2 * 36) * 150 = 50 000 credits

> Charging rates
> --------------

Reviewer: ```suggestion
Charge rates
------------
```

The charge rate for the various node types of Genius and wICE are listed in the table
below.  
The reported cost is the number of Slurm credits needed per core/GPU per minute.

> +---------+-----------------+------------------------+

Author: Originally, this page contained how to calculate the required credits per each job. Once I converted the content to Slurm, then I had to mention the weights in the formula. Consequently, I also had to list the weights, too to complement the whole text, to show how many Slurm credits are needed in a job. That's why it has appeared here. 

The service catalog has a slightly different focus; it tells how much Slurm credits and money (in EUR) costs per using different resources. Then users can convert their pocket money into Slurm credits before filling in the credit request form.

Given the fact that these two information complement one another, the link to the service catalog is also provided on `credits.rst` page. 

With this, I think users can have all necessary information, each from a relevant resource with minimal overlap in content.

What do you think?

> +---------+-----------------+------------------------+

Reviewer 2: My point is that these values are harder to interpret than what is in the service catalog, so they add little value since users can retrieve them themselves with the command shown earlier on this page. It is not necessarily bad that this table is included, but of course it has to be maintained when we update values or add a partition (for instance superdome).

Job cost calculation
~~~~~~~~~~~~~~~~~~~~

On Tier-2 clusters, we use Slurm for acounting purposes (on top of resource and 
job management).
See :ref:`Slurm accounting <accounting_leuven>` page for additional information.
In Slurm terminology, the cost of a job depends on the trackable resources (TRES)
it consumes. Two distinct TRES are the number of CPU cores and GPU devices. 
Furthermore, different CPU and GPU architectures are weighted by a constant factor
according to their generation.

In order to retrieve the configured weights (precisely speaking ``TRESBillingWeights``) 

> e.g. on wICE, you can execute::

Author: I realized (`grep`) that I am hardly using "architecture" in the text. But, indeed, "node type" is a better suited word.

**CPU-only jobs**

The following formula applies::

   CPU TRESBillingWeights \* num_cores \* walltime

Where

- ``CPU TRESBillingWeights`` is the applied weight for CPU resources (see above)
- ``num_cores`` is the *effective* number of cores used for the job
- ``walltime`` is minutes that the job consumed the resources

**GPU jobs**

The following formula applies::

   (CPU TRESBillingWeights \* num_cores + GPU TRESBillingWeights \* num_gpus) \* walltime
   
Where

- ``CPU TRESBillingWeights`` is the applied weight for CPU resources (see above)
- ``GPU TRESBillingWeights`` is the applied weight for GPU resources (see above)
- ``num_cores`` is the *effective* number of cores used for the job
- ``num_gpus`` is the *effective* number of GPUs used for the job
- ``walltime`` is minutes that the job consumed the resources

.. note::

    *effective* number of cores is not necessarily equal to what the user requests.
    E.g. if a job requests a single core/task, but the full memory of a node on wICE,
    then one node is blocked for such a job. Consequently, the effective number of cores
    will be 72, instead of 1.

>     Similarly, on a node with 4 GPUs, if a user requests to use a single GPU, one-forth

Author: fixed via `d13094eb9`

    `HPC Service Catalog<https://icts.kuleuven.be/sc/onderzoeksgegevens/hpc_vsc_page>`
    (login required).

The difference in cost between different machines/processors reflects
the performance difference between those types of nodes. The total cost
of a job will typically be the same on any compute nodes, but the
walltime will be different, depending on the performance of the nodes.

In the examples below, you run your jobs on a ``skylake`` node, for which
we charge 10 000 Slurm credits per hour.

An example of a job running on multiple nodes and cores is given below::

   $ sbatch -A lp_astrophysics_014 -M genius -N 2 -ntasks-per-node 36 simulation_3415.slurm

For Genius thin nodes we have ``TRESBillingWeights=CPU=4.62963``.
If this job finishes in 2.5 hours (i.e., walltime is 150 minutes), the user
will be charged::

   4.62963 \* (2 \* 36) \* 150 = 50 000 credits

How do I get credits to use the Tier-1 infrastructure
-----------------------------------------------------

Access to the Tier-1 is project-based, if you have a starting grant or
an approved project, or you pay for your compute time, you should have
received information on your job credits.

> If not, please refer to the `official VSC website <https://www.vscentrum.be/>`, or

Author: fixed via `f363878381f`
